[ADD] member_card: add member card report

tfrancoi opened the pull request:
This module add the ability to print the member card
from the contact to be use with beesdoo_base

tfrancoi commented:
TODO: Remove polln specific logo
Add a field on the res.company to store the logo on the card

robinkeunen reviewed (approved):
added readme and todo

carmenbianca commented:
I'm a little confused. This PR adds `polln_base`, not `member_card`.

robinkeunen commented:
Indeed, that's why it is labelled with "TODO" : thibault still needs to extract the member card functionnalities to a new module out of polln_base.

polchampion commented:
@remytms Est-ce que tu pourras estimer ceci (pas urgent) ? (avec le todo logo)

Virg et Pol : option pour l'intercoop, besoin d'une estimation pour avancer

victor-champonnois commented:
Note : I've extracted member_card from beesdoo_base in this PR https://github.com/beescoop/Obeesdoo/pull/384 
@polchampion I can work on adding this member_card report on top of the beesdoo_base split if that's OK

victor-champonnois commented:
~~@polchampion is this tested and OK to merge ?~~

robinkeunen committed:
[REF] polln_base: pre-commit

tfrancoi committed:
[ADD] member_card: add member card report

This module add the ability to print the member card
from the contact to be use with beesdoo_base

victor-champonnois commented:
branch rebased

victor-champonnois commented:
This conflicts with https://github.com/beescoop/Obeesdoo/pull/384

robinkeunen commented:
work was started here : https://github.com/beescoop/Obeesdoo/pull/432

victor-champonnois committed:
[FIX] rem duplication w/ eater_member_card

victor-champonnois committed:
[REF] move polln_base's pollncard to member_card

victor-champonnois committed:
[ADD] configurable member_card logo

victor-champonnois committed:
[REF] specific template for eater_member_card

victor-champonnois commented:
- integrated the functionality in `member_card` and `eater_member_card` module. 
- The report in `eater_member_card` replaces the one in `member_card`, because I did not find a way to inherit it and add the `partner.parent_eater_id.name`, because of the foreach loop. Cf this [discussion](https://framateam.org/coopiteasy/pl/4gqyq6jswtrg3czwq5g3477pjh) and this [comment](https://github.com/beescoop/Obeesdoo/blob/d107373cf25ded5868688d0c5684acf778fd7f4c/eater_member_card/report/member_card_template.xml#L12)
- the commit [dc2bc07](https://github.com/beescoop/Obeesdoo/pull/289/commits/dc2bc07c93e91fbc95eb63ab77e11dfcf7507f1e) is not related to the task, it is a quick fix en passant.

The [Odoo task ](https://gestion.coopiteasy.be/web#id=9089&model=project.task&view_type=form&menu_id=)

victor-champonnois committed:
[FIX] limit name to 20 char to avoid affecting the card size

robinkeunen commented on eater_member_card/report/member_card_template.xml:
```diff
@@ -0,0 +1,64 @@
+
+
+
+
```
too many blank lines

victor-champonnois replied:
fixed

robinkeunen commented on member_card/readme/DESCRIPTION.rst:
```diff
@@ -6,9 +6,8 @@ Create a member card and link it to a partner.
 - the partner's card and barcode history is visible in the member card tab
 - the barcode of a partner is computed from the last active member card
 - a boolean "Print Member card?" allows to filter on partners for which you need to print new cards.
+- contain a template for the member card
```
```suggestion
- contains a generic template for the member card
```

robinkeunen commented on eater_member_card/report/member_card_template.xml:
```diff
@@ -0,0 +1,64 @@
+
+
+    <template id="member_card_template">
+        <!-- replace the member_card template to add worker information.
+            Cannot use template inheritance because the worker information is in partner variable,
+        which is set in the foreach loop -->
```
I'm surprised by this assertion, it works in other report inheritance 🤔

victor-champonnois replied:
I could not make it work, do you have an example ?

victor-champonnois replied:
@robinkeunen up :)

victor-champonnois committed:
Update member_card/readme/DESCRIPTION.rst

Co-authored-by: Robin Keunen <robin@keunen.net>

robinkeunen committed:
[IMP] member_card: remove dependency to partner_firstname

robinkeunen committed:
[IMP] member_card: tests

robinkeunen committed:
[IMP] eater_member_card: move parent_barcode to new module

robinkeunen committed:
[FIX] eater_member_card: fix template inheritance

robinkeunen commented:
@victor-champonnois can you have a look at my changes ? J'ai un peu craqué mon slip :o

robinkeunen commented:
@polchampion this was deployed on test server. Can you test it functionally  ?

`member_card`
```
Create a member card and link it to a partner.

- each member card has a barcode
- create new cards in the "member card" tab of the partner form
- use "Force Barcode" field to set a user defined barcode
- the partner's card and barcode history is visible in the member card tab
- the barcode of a partner is computed from the last active member card
- a boolean "Print Member card?" allows to filter on partners for which you need to print new cards.
- contains a generic template for the member card
- the card template uses field `member_card_logo` rather than the company logo

The wizards "Request member card printing" and "Set member card as printed" allow to

Careful : this module overrides the barcodes already defined on the partners.

If point of sale is installed, the generated barcode matches customer pattern rule.
```

`eater_member_card`
```
- eater barcode is computed from parent's barcode.
- Print Member card? is set to True when a new eater is created.
```

`eater_parent_code`
```
Adds field parent_barcode. This legacy field is necessary to generate
member card report with Jaspersoft.

This module depends on eater_member_card to generate the partner barcodes.

Do not migrate to 16.0
```

victor-champonnois commented on eater_member_card/models/partner.py:
```diff
@@ -1,31 +1,28 @@
-                for c in partner.member_card_ids:
-                    if c.valid:
-                        partner.barcode = c.barcode
+        parents = self.filtered(lambda p: not p.parent_eater_id)
+        eaters = self.filtered(lambda p: p.parent_eater_id)
+        super(Partner, parents)._compute_bar_code()
+
+        # warning : there is not constraint checking that parent_eater_id
+        # is set if eater field is set to eater
+        for partner in eaters:
+            partner.barcode = partner.parent_eater_id.barcode
```
```suggestion
        parents = self.filtered(lambda p: not p.parent_eater_id)
        super(Partner, parents)._compute_bar_code()
        
        eaters = self.filtered(lambda p: p.parent_eater_id)
        # warning : there is not constraint checking that parent_eater_id
        # is set if eater field is set to eater
        for partner in eaters:
            partner.barcode = partner.parent_eater_id.barcode
```
Just reordering lines to make the operations clearer

robinkeunen replied:
👍

victor-champonnois reviewed (approved):
Thanks for the changes :) 
@polchampion Virginie is the reference analyst on this task (cf [task](https://gestion.coopiteasy.be/web#id=9089&action=479&model=project.task&view_type=form&menu_id=536)), I ask her to test.

victor-champonnois committed:
Merge branch '12.0' into 12.0-member-card

victor-champonnois commented:
/ocabot rebase
